[Feat] 캘린더 생성 & 편집 UI 구현

zero0205 opened the pull request:
## #️⃣ 연관된 이슈

- #20

## 📋 작업 요약

shadcn/ui의 `Dialog`와 react hook form을 사용하여 캘린더 생성 UI와 캘린더 편집 UI 구현

|![image](https://github.com/user-attachments/assets/23fb1c14-9ead-4a2a-8217-449a3c348a7b)|![image](https://github.com/user-attachments/assets/bf051ba4-6f85-406d-aead-afad9837a84f)|
|--|--|
|캘린더 생성|캘린더 편집|
|![image](https://github.com/user-attachments/assets/907a740f-6614-4e48-b088-9cf042f39631)||
|멤버 UI 최대 크기(스크롤)||


## 🔍 세부 내용

### 캘린더 생성

- shadcn/ui의 `Dialog`를 사용하여 모달 형태로 구현
- react hook form을 사용한 입력 폼 관리
- react hook form의 `useFieldArray`를 사용하여 같은 필드에 여러 개의 데이터가 들어오는 경우(participants) 관리

### 캘린더 편집

- shadcn/ui의 `Dialog`를 사용하여 모달 형태로 구현
- react hook form을 사용한 입력 폼 관리
- `useRef`와 `useState`를 사용하여 participants 입력 및 상태 관리
  - 캘린더 생성에서는 입력 폼의 데이터를 모두 모아서 한 번에 API 요청을 하기 때문에 participants를 form 데이터로 취급했지만, 캘린더 편집에서는 캘린더의 이름과 닉네임을 수정하는 API와 멤버 초대 API가 별도로 존재하기 때문에 participants를 별도의 state로 관리.

### entities 레이어에서 타입 수정

- entities/calendar 내의 calendarParticipant 타입 수정
- entities/member 슬라이스 생성(타입 정의)

## 🤔 고민했던 내용

### 캘린더 생성과 편집의 공통 컴포넌트 사용?

처음에는 CalendarForm이라는 공통 컴포넌트를 두고, 이 공통 컴포넌트를 CreateCalendar와 EditCalendar에서 가져다 쓰는 방식으로 구현하려고 했다. 그렇지만 구현하다보니 둘의 동작 방식이 너무 달랐다.

- `CreateCalendar`: 캘린더 정보 + 멤버 목록을 한 번에 서버로 전송
- `EditCalendar`: 캘린더 정보(캘린더 이름, 닉네임) 수정과 멤버 초대를 분리된 API로 처리

물론 공통 컴포넌트로 구현한다면 가능은 하겠지만, 다음과 같은 문제가 발생할 수 있다.

- props와 내부 로직이 과하게 복잡해짐
- 조건부 렌더링과 분기 처리로 인한 가독성 저하

이런 이유로 CreateCalendar와 EditCalendar는 비슷한 UI를 갖고 있더라도 공통 컴포넌트를 사용하지 않고, 서로 독립된 컴포넌트로 구현했다. 

다만, 추후에 개선한다면 작은 컴포넌트로 나누어서 공통으로 사용하는 구조로 개선은 기대해볼 수 있을 것 같다.

zero0205 committed:
:sparkles: [Feat]: 캘린더 생성/수정 폼 컴포넌트 \& 캘린더 생성 UI 구현

- entities/member 추가(타입 정의)
- entities/calendar의 participant 관련 타입 수정
- shadcn/ui의 Dialog 컴포넌트와 react hook form을 사용하여 캘린더 생성/수정 폼 구현

zero0205 committed:
:recycle: [Refactor]: DialogDescription이 없음을 명시적으로 표시

- DialogContent에 aria-describedby={undefined}을 추가하여 추가 설명이 없음을 명시적으로 표시
- 웹 접근성 관련 이슈 수정

zero0205 committed:
:bug: [Fix]: Dialog가 자동으로 닫히는 문제 수정

- 마우스가 브라우저를 벗어나면 Dialog가 닫히는 현상 수정
- CreateCalendar 컴포넌트가 조건부 렌더링 되고 있던 것이 원인

zero0205 committed:
:lipstick: [Design]: 참여자 숫자에 따라 Dialog 크기 동적 조절

zero0205 committed:
:sparkles: [Feat]: 캘린더가 닫힐 때 초기화되도록 수정

- X 버튼을 누르거나 Dialog 외부를 클릭하여 Dialog가 닫힐 때, 폼이 초기화되도록 수정
- CalendarFormDialog를 사용할 때, key props를 넣어서 컴포넌트를 강제로 재마운트

zero0205 committed:
:sparkles: [Feat]: 캘린더 생성 컴포넌트 분리

- 캘린더 생성과 수정을 하나의 폼으로 처리하려던 기존의 구조는 너무 복잡도가 높다고 판단되어서 별도의 features 슬라이스로 분리

zero0205 committed:
:sparkles: [Feat]: 캘린더 편집 UI 구현

seungheon123 reviewed (commented):
캘린더 생성/편집 창에 생각보다 공백이 많네요...😅 혹시 창 사이즈 줄여보는거 어떨까요?

zero0205 committed:
:lipstick: [Design]: 캘린더 멤버 UI의 높이 동적으로 늘어나도록 수정

zero0205 commented:
@seungheon123 

멤버 수에 따라 동적으로 늘어나도록 수정했습니다! PR 내용도 수정해놨습니다😃

zero0205 committed:
:lipstick: [Design]: 멤버의 nickname이나 email이 길어질 경우 truncate 처리

zero0205 committed:
:sparkles: [Feat]: 캘린더 삭제 시 캘린더 이름 입력 후 삭제 로직 구현

- 정해진 텍스트 입력하여 확인 후 로직을 수행할 수 있는 컴포넌트(TextConfirmDialog) 구현
- EditCalendar에서 캘린더 삭제 버튼 부분에 TextConfirmDialog 사용

zero0205 committed:
:sparkles: [Feat]: 캘린더 편집에서 멤버 삭제 시 확인 Dialog 추가

- 확인 Dialog(ConfirmDialog) 구현
- 캘린더 편집의 멤버 삭제 버튼에 ConfirmDialog 추가

zero0205 commented:
/gemini review

zero0205 committed:
:sparkles: [Feat]: Shared/ui의 공개 API 수정

- ConfirmDialog와 TextConfirmDialog도 export
